optimade/validator/validator.py

```diff
@@ -171,63 +184,98 @@ def test_case(test_fn):
     @wraps(test_fn)
     def wrapper(validator, *args, optional=False, **kwargs):
         try:
-            result, msg = test_fn(validator, *args, **kwargs)
-        except json.JSONDecodeError as exc:
-            result = None
-            msg = (
-                "Critical: unable to parse server response as JSON. "
-                f"Error: {type(exc).__name__}: {exc}"
-            )
-        except (ResponseError, ValidationError) as exc:
-            result = None
-            msg = f"{type(exc).__name__}: {exc}"
-
-        try:
-            request = validator.client.last_request
-        except AttributeError:
-            request = validator.base_url
+            try:
+                result, msg = test_fn(validator, *args, **kwargs)
+            except json.JSONDecodeError as exc:
+                msg = (
+                    "Critical: unable to parse server response as JSON. "
+                    f"Error: {exc.__class__.__name__}: {exc}"
+                )
+                raise exc
+            except (ResponseError, ValidationError) as exc:
+                msg = f"{exc.__class__.__name__}: {exc}"
+                raise exc
+            except Exception as exc:
+                msg = f"{exc.__class__.__name__}: {exc}"
+                raise InternalError(msg)
+
+        except Exception as exc:
+            result = exc
+            traceback = tb.format_exc()
+
+        finally:
+            try:
+                request = validator.client.last_request
+            except AttributeError:
+                request = validator.base_url
 
-        if result is not None:
-            if not optional:
-                validator.success_count += 1
-            else:
-                validator.optional_success_count += 1
-            message = f"✔: {request} - {msg}"
-            if validator.verbosity > 0:
-                if optional:
-                    print(message)
+            if not isinstance(result, Exception):
+                if not optional:
+                    validator.success_count += 1
                 else:
-                    print_success(message)
-            else:
-                if optional:
+                    validator.optional_success_count += 1
+                message = f"✔: {request} - {msg}"
+                if validator.verbosity > 0:
```

Comment:
If you prefer the more explicit version, be my guest :)
```suggestion
                if validator.verbosity:
```